scipy/signal/spectral.py

```diff
@@ -511,11 +511,406 @@ def spectrogram(x, fs=1.0, window=('tukey',.25), nperseg=256, noverlap=None,
+        Specifies whether the input signal was padded via odd extension, by
+        using ``centered=True`` in `stft`. Defaults to True.
+    strict : bool, optional
+        Specifies whether a ``ValueError`` will be raised if the choice of
+        window and overlap does not satisfy the COLA constraint. If ``False``,
+        execution will proceed uninterrupted, but faithful reconstruction may
+        not occur.
+    time_axis : int, optional
+        Where the time segments of the STFT is located; the default is the
+        last axis (i.e. ``axis=-1``).
+    freq_axis : int, optional
+        Where the frequency axis of the STFT is located; the default is the
+        penultimate axis (i.e. ``axis=-2``).
+
+    Returns
+    -------
+    t : ndarray
+        Array of output data times.
+    x : ndarray
+        iSTFT of Zxx.
+
+    See Also
+    --------
+    stft: Short Time Fourier Transform
+    check_COLA: Check whether the Constant OverLap Add (COLA) constraint is met
+
+    Notes
+    -----
+    In order to enable inversion of an STFT via the inverse STFT in
+    `istft`, the signal windowing must obey the constraint of "Constant
+    OverLap Add" (COLA). This ensures that every point in the input data is
+    equally weighted, thereby avoiding aliasing and allowing full
+    reconstruction. Whether a choice of `window`, `nperseg`, and `noverlap`
+    satisfy this constraint can be tested with `check_COLA`.
+
+    .. versionadded:: 0.18.0
+
+    References
+    ----------
+    .. [1] Oppenheim, Alan V., Ronald W. Schafer, John R. Buck
+           "Discrete-Time Signal Processing", Prentice Hall, 1999.
+    .. [2] Daniel W. Griffin, Jae S. Limdt "Signal Estimation from Modified
+           Short Fourier Transform", IEEE 1984, 10.1109/TASSP.1984.1164317
+
+    Examples
+    --------
+    >>> from scipy import signal
+    >>> import matplotlib.pyplot as plt
+
+    Generate a test signal, a 2 Vrms sine wave corrupted by 0.001 V**2/Hz of
+    white noise sampled at 1024 Hz.
+
+    >>> fs = 1024
+    >>> N = 10*fs
+    >>> amp = 2 * np.sqrt(2)
+    >>> noise_power = 0.001 * fs / 2
+    >>> time = np.arange(N) / float(fs)
+    >>> carrier = amp * np.sin(2*np.pi*2e3*time)
+    >>> noise = np.random.normal(scale=np.sqrt(noise_power), size=time.shape)
+    >>> x = carrier + noise
+
+    Compute the STFT, and zero the components that are 10% or less of the
+    carrier magnitude.
+
+    >>> f, t, Zxx = signal.stft(x, fs, centered=True)
+    >>> Zxx = np.where(np.abs(Zxx) >= amp/10, Zxx, 0)
+
+    Convert back to a time series via inverse STFT
+
+    >>> _, xrec = signal.istft(Zxx, fs, centered=True)
+
+    Compare the cleaned signal with the original and true carrier signals.
+
+    >>> plt.plot(time, x, time, xrec.real, time, carrier)
+    >>> plt.xlim([0, 0.1])
+    >>> plt.xlabel('Time [sec]')
+    >>> plt.ylabel('Signal')
+    >>> plt.legend(['Carrier + Noise', 'Filtered via STFT', 'True Carrier'])
+    >>> plt.show()
+
```

Comment:
The fix for stft looks good, but there is something strange here as well.
If I try to plot the STFT, I would expect a frequency of 2kHz
 ```
   f = fftshift(f)
    plt.pcolormesh(t, f, np.abs(fftshift(Zxx, axes=0)))  
    plt.ylim([0, f[-1]])
    plt.title('STFT Magnitude')
    plt.ylabel('Frequency [Hz]')
    plt.xlabel('Time [sec]')
    plt.show()
```
but the plot shows only 50Hz, though I guess it would be impossible to see 3kHz, as the time resolution is only ms?
It would be informative to add a view of the STFT here as well.